### PR TITLE
[WIP] Add center-based truth definition, update hinge

### PR DESCRIPTION
**Pull request by lgray**
This PR will allow us to use the centers-based hinge truth definition.

I'll also add in the switch to a flat / unscheduled learning rate.

We can optimize the LR scheduling once we're happy with the number of tracks we can predict and ability to generalize.

### CODE REVIEW

**Comment by lgray** [comment]
Made a fully tensor op version of generating the truth based on centers:
```python
def center_embedding_truth_tensor(coords, truth_label_by_hits, device='cpu'):
    truth_ordering = torch.argsort(truth_label_by_hits)    
    uniques, counts = torch.unique(truth_label_by_hits, return_counts=True)

    n_hits = truth_label_by_hits.size()[0]
    n_clusters = uniques.size()[0]
    
    centers = scatter_mean(coords, truth_label_by_hits, dim=0, dim_size=n_clusters)

    all_dists = torch.cdist(coords.expand(n_clusters,-1,-1).contiguous(), centers[:,None])
    copied_truth = truth_label_by_hits.expand(n_clusters,-1)
    copied_uniques = uniques[:,None].expand(-1, n_hits)
    truth = torch.where(copied_truth == copied_uniques,
                        torch.ones((n_clusters, n_hits), dtype=torch.int64, device=device), 
                        torch.full((n_clusters, n_hits), -1, dtype=torch.int64, device=device))
    
    dists_out = all_dists.reshape(n_clusters*n_hits)
    truth_out = truth.reshape(n_clusters*n_hits)
    
    return (dists_out, truth_out)
```

but it's not compatible with the old pairwise truth. I'll just switch everything to this if it's OK?

**Comment by lgray** [comment]
Numbafied version of cluster matcher.
```python3
import numba as nb

@nb.njit()
def build_cluster_list(indices, clusters, labels):
    pred_clusters = []
    for label in labels:
        pred_clusters.append(indices[clusters == label])
    return pred_clusters

@nb.njit()
def do_matching(indices, pred_labels, true_labels, data_y, data_y_particle_barcodes, data_truth_barcodes, 
               data_truth_pt, data_truth_eta, data_truth_phi):
    matched_pred_clusters = []
    true_cluster_properties = np.zeros((len(true_labels), 3), dtype=np.float32)
    for i, label in enumerate(true_labels):
        true_indices = set(indices[data_y == label])
        best_pred_cluster = -1
        best_iou = 0
        for j, pc in enumerate(pred_labels):
            #print('i-pred',i)
            pc = set(pc)
            isec = true_indices & pc
            iun = true_indices | pc
            iou = len(isec)/len(iun)
            if best_pred_cluster == -1 or iou > best_iou:
                best_pred_cluster = j
                best_iou = iou
        matched_pred_clusters.append(best_pred_cluster)
        # now make the properties vector
        thebc = np.unique(data_y_particle_barcodes[data_y == label])[0]
        select_truth = (data_truth_barcodes == thebc)
        pt_inv = np.reciprocal(data_truth_pt[select_truth])
        eta = data_truth_eta[select_truth]
        phi = data_truth_phi[select_truth]
        true_cluster_properties[i][0] = pt_inv[0]
        true_cluster_properties[i][1] = eta[0]
        true_cluster_properties[i][2] = phi[0]

    return matched_pred_clusters, true_cluster_properties

def match_cluster_targets(clusters, truth_clusters, data):
    np_truth_clusters = truth_clusters.cpu().numpy()
    true_cluster_labels = np.unique(np_truth_clusters)   
    np_clusters = clusters.cpu().numpy()
    pred_cluster_labels = np.unique(np_clusters)
    pred_cluster_mask = np.ones_like(np_truth_clusters, dtype=np.bool)
        
    indices = np.arange(np_clusters.size, dtype=np.int64)
    pred_clusters = build_cluster_list(indices, np_clusters, pred_cluster_labels)
    
    matched_pred_clusters, true_cluster_properties = \
        do_matching(indices, pred_clusters, true_cluster_labels, 
                    data.y.cpu().numpy(),
                    data.y_particle_barcodes.cpu().numpy(),
                    data.truth_barcodes.cpu().numpy(),
                    data.truth_pt.cpu().numpy(), 
                    data.truth_eta.cpu().numpy(), 
                    data.truth_phi.cpu().numpy())

    matched_pred_clusters = np.array(matched_pred_clusters, dtype=np.int64)
    pred_indices = torch.from_numpy(matched_pred_clusters).to(clusters.device)
    
    true_cluster_properties = np.array(true_cluster_properties, dtype=np.float)
    y_properties = torch.from_numpy(true_cluster_properties).to(clusters.device).float()
    
    return pred_indices, y_properties
```

**Comment by lgray** [comment]
New code is in. I will test it once this training run is over. Later today!

**Review by chhavisharma — APPROVED** [review]
